FEATURE: Add NodeLink fusion prototype

gradinarufelix opened the pull request:
Add the ``Neos.Neos:NodeLink`` fusion prototype which renders an ``<a>`` tag based on ``Neos.Neos:NodeUri``, similar to ``Neos.Neos:ImageTag`` and ``Neos.Neos:ImageUri``. 

- all properties of ``Neos.Neos:NodeUri`` are supported and passed over
- ``attributes`` (:ref:`Neos_Fusion__Attributes`) Link tag attributes
- ``content`` (string) The label of the link, defaults to the node label ``q(node).property('_label')``.

This helps reducing code lines especially when using fusion-afx, with this prototype you can now directly link nodes in an afx renderer.

mficzel commented on Neos.Neos/Resources/Private/Fusion/Prototypes/NodeLink.fusion:
```diff
@@ -0,0 +1,25 @@
+# Renders an <a> tag based on Neos.Neos:NodeUri pointing to a node
+#
+prototype(Neos.Neos:NodeLink) < prototype(Neos.Fusion:Tag) {
+  node = 'pass-the-node'
```
Please do not define a string default value where a NodeInterface is required, i suggest to use `null` as default.

gradinarufelix replied:
Okay, I also found that weird. But in ImageUri it is also solved like that, which makes absolutely no sense if we expect an object. I'll change that.

mficzel commented on Neos.Neos/Resources/Private/Fusion/Prototypes/NodeLink.fusion:
```diff
@@ -0,0 +1,25 @@
+    }
+  }
+
+  content = ''
```
Maybe we could even define the default content as `${q(this.node).property('_label')}` instead of an empty title. I am not totally sure wether this is actually a good idea.

jonnitto replied:
This would be a menu-like behavior, then. Why not?

gradinarufelix replied:
Better than an empty string probably, yes.

jonnitto commented:
First, thanks @gradinarufelix for you PR!!

I don't know exactly if I should support this prototype. Not because I don't see the need, but I experienced that there are often much more requirements for this type of prototypes: 
* Should the link be hidden if there is no node set?
* Or just in the frontend?
* Or switching to a `<span>` in the frontend?

All because of that, I wrote my own prototypes who takes care of all these things. Like that, I can reuse it over and over again and improve it the way I need. On the other hand, it is really the same with `ImageTag` and `ImageUri`… (I never use `ImageTag`, because I miss there many thinks like `srcset` and so on)

So: No Up or downvoting from me, I just want to share my thoughts…

gradinarufelix commented:
> First, thanks @gradinarufelix for you PR!!
> 
> I don't know exactly if I should support this prototype. Not because I don't see the need, but I experienced that there are often much more requirements for this type of prototypes:
> 
> * Should the link be hidden if there is no node set?
> * Or just in the frontend?
> * Or switching to a `<span>` in the frontend?
> 
> All because of that, I wrote my own prototypes who takes care of all these things. Like that, I can reuse it over and over again and improve it the way I need. On the other hand, it is really the same with `ImageTag` and `ImageUri`… (I never use `ImageTag`, because I miss there many thinks like `srcset` and so on)
> 
> So: No Up or downvoting from me, I just want to share my thoughts…

Thanks for your thoughts @jonnitto ! Yeah, we also had that discussion when talking about the new fusion-based menu. But from my point of view we should have some basic prototypes in the package as well as an entry point for newbies. I learned a lot from them for example. 

Of course if you are a power-user you will write your own prototypes, but I think they are not conflicting at all. 

I'll think about that three points and maybe suggest a solution.

mficzel commented:
To add more Background: This is mainly for easier NodeLinking in AFX which was a bit problematic until now: 

```
afx`
    <Neos.Neos:NodeLink node={props.node} />
`
```

This is not needed in pure presentational-fusion where i would always pass uris instead of nodes but for menus similar stuff this is very usefull.

jonnitto commented:
That's correct. For menus it is useful. @mficzel What do you think? Would you add a `@if.hasNode = ${this.node}` or `@if.hasNodeOrInBackend = ${this.node || node.context.inBackend}` to it?

mficzel commented:
@jonnitto I am not sure what the expected behaviour is if `node` is missing. I think the `@if.hasNodeOrInBackend` rule is already too special and you have to look into code to understand and override it. So i suggest to render the tag always and let integrator add the needed @if statements. 

BTW: Those arguments are valid for the default `content`, but since this is overwritten very easy i would accept  `content = ${q(this.node).property('_label')}` that is suggested myself.

jonnitto commented:
I've set the default values as we said. From my point of view, it is ready to merge.

dimaip commented:
I would really love to get a review from @kitsunet before merging this one.
Personally I'm somehow intuitively against such helper prototypes in core, but I don't have enough energy left to explain property.

gradinarufelix commented:
I've added also addQueryString, absolute and baseNodeName as they are also arguments of NodeUri and I added the documentation. But I'm still unsure about the docs, neved worked with that before. Please have a look.

kitsunet commented:
I am also a bit torn like @jonnitto about if we really need it or not, but I agree that you can just ignore it and do your own thing so it doesn't really hurt the power users. Therefore generally +1 to the idea. I would expect it to be component though, now that we have components in core. Would avoid all these nasty `@context` things :)

gradinarufelix commented:
I talked about the component thing with @mficzel and we agreed to not do it as a component for the time. But if we change everything to components anyway then yeah, why not.

mficzel commented:
I am for adding this because since we say that afx is best practice we should make all common tasks easy in afx. Linking nodes is one of first things people do and this should not need advanced concepts like @path and @context.

kitsunet commented:
I am unclear what speaks against a component @mficzel ? I am still very strongly opposed to `@context.something = ${this.something}` because it's very bad and we shouldn't do it (because it suggests it's a good way to do things).

mficzel commented:
@kitsunet i would start refactoring core fusion to components with a major. That is the only reason why i would still add this the oldscool way.

jonnitto commented on Neos.Neos/Documentation/References/NeosFusionReference.rst:
```diff
@@ -1023,6 +1023,24 @@ Example::
+
+:\*: All :ref:`Neos_Neos__NodeUri` properties
+:attributes: (:ref:`Neos_Fusion__Attributes`) Link tag attributes
+:content: (string) The label of the link.
```
I'm missing here the description of the default value

mficzel replied:
Correct, is adjusted

bwaidelich commented on Neos.Neos/Resources/Private/Fusion/Prototypes/NodeLink.fusion:
```diff
@@ -0,0 +1,34 @@
+  @context.argumentsToBeExcludedFromQueryString = ${this.argumentsToBeExcludedFromQueryString}
+  @context.addQueryString = ${this.addQueryString}
+  @context.absolute = ${this.absolute}
+  @context.baseNodeName = ${this.baseNodeName}
```
nitpick: Readability could be improved by putting this into one `@context {...}` block IMO

bwaidelich commented on Neos.Neos/Resources/Private/Fusion/Prototypes/NodeLink.fusion:
```diff
@@ -0,0 +1,34 @@
+# Renders an <a> tag based on Neos.Neos:NodeUri pointing to a node
```
```suggestion
# Renders an <a> tag based on Neos.Neos:NodeUri pointing to a node
# The link text is the node label, unless overridden
```

mficzel replied:
Is adjusted

bwaidelich commented on Neos.Neos/Resources/Private/Fusion/Prototypes/NodeLink.fusion:
```diff
@@ -0,0 +1,34 @@
+    }
+  }
+
+  content = ${q(this.node).property('_label')}
```
is `_label` an "official" property for the node's label?

Sebobo replied:
Should be just `node.label`

Sebobo commented:
I would actually like to have the title attribute be rendered automatically. People can still override it for customization. I know that having the title the same as the content doesn't bring value, but then I would only need to override content in most cases and in other cases where nobody knows what to put in there it would at least quiet down all the missing attribute warnings ;-)

Rest looks good and would be quite useful IMHO. Just use `node.label` for the content would be better.

bwaidelich commented on Neos.Neos/Documentation/References/NeosFusionReference.rst:
```diff
@@ -1023,6 +1023,24 @@ Example::
+
+:\*: All :ref:`Neos_Neos__NodeUri` properties
+:attributes: (:ref:`Neos_Fusion__Attributes`) Link tag attributes
+:content: (string) The label of the link, defaults to the node label ``q(node).property('_label')``.
```
```suggestion
:content: (string) The label of the link, defaults to the node label ``node.label``.
```

mficzel replied:
Is adjusted

bwaidelich commented on Neos.Neos/Resources/Private/Fusion/Prototypes/NodeLink.fusion:
```diff
@@ -0,0 +1,34 @@
+    }
+  }
+
+  content = ${q(this.node).property('_label')}
```
```suggestion
  content = ${this.node.label}
```
like so?

mficzel replied:
Is adjusted to ``node.label`` as ``this.node`` was put into ``@context.node`` anyways.

mficzel commented:
@bwaidelich @Sebobo i adjusted the pr as you suggested. Please re review.
